# RCPA WGS PTP Bioinformatics Pipeline

## Background

In this repository, we document and share the Bioinformatics pipeline used to analyse 
FASTQ data generated by Australian public health laboratories participating in 
the RCPA's whole-genome sequencing (WGS) proficiency testing program (PTP).

## Running the pipeline

### Software Dependencies

* Python >= 3.6
* Snakemake >= 5.4
* Singularity >=3.1

### Hardware requirements

* 8 GB of RAM (16 GB is recommended if running within a Virtual Machine)
* 30 GB of Hard Disk space

### Clone the repo

Clone the repository:

```console
git clone https://github.com/andersgs/rcpa_wgs_ptp.git
```

#### Running on Unix


```console
cd rcpa_wgs_ptp
bash startup.sh
snakemake --use-singularity # run the test example
```

`bash startup.sh` will download the singularity images, and a test dataset.

Then, run `snakemake --use-singularity` will run the pipeline.

To run it with your own data, follow instructions below.

### Mac

```console
cd rcpa_wgs_ptp
vagrant up
vagrant ssh
cd qc_folder && snakemake --use-singularity
```

To install `vagrant` and `virtualbox`. I suggest using `brew`:

`brew cask install vagrant virtualbox`

Then:

`vagrant up` --- will bootstrap the Ubuntu box with Singularity installed, and will download all the Singularity images.

`vagrant ssh` --- will log you on to the Ubuntu virtual machine.

`cd qc_folder && snakemake --use-singularity` --- will run the test dataset

## Tools

1. Read stats
   1. `seqtk`
2. Read trimming
   1. `timmomatic`
3. Estimating genome size based on kmers
   1. `mash`
4. Species identification
   1. `kraken2` --- for the purposes of the WGS PTP we are only using the 8GB minikraken DB.
5. Assembly
   1. `shovill` --- using `spades 3.13.0`
6. Sequence Typing
   1. `mlst` --- using schemes available on PubMLST
7. Serotyping
   1. `SISTR` for *Salmonella*
8. AMR detection
   1. `abricate` --- using the NCBI database